### PR TITLE
(#119-3)To make it look a bit more receipt like, table with EIN

### PR DESCRIPTION
**Pull request by ronaldblanco**
#119 -3 To make it look a bit more receipt like, there could be another table as part of the donation receipt with the organization details and EIN number from general settings.
Solution tested locally and checked with lint command.

### CODE REVIEW

**Comment by jspaine** [comment]
I meant it should be part of the receipt placeholder. You can't edit tables with Quill and I'm not sure what'll happen if you insert one into the editor like that. What does it look like!?

**Comment by ronaldblanco** [comment]
In the console.log looks well (the same way that the other)

**Comment by jspaine** [comment]
I mean what does it look like in the email editor?

I think it should be in the same placeholder as the receipt.

**Comment by ronaldblanco** [comment]
I will be checking.

**Comment by ronaldblanco** [comment]
Hey @jspaine check now, I did create a new placeholder I think it is a better solution (as I understand)
Let me know.

Thanks.

File: common/placeholders.js [hunk 1]
```diff
@@ -10,6 +10,7 @@ const placeholders = [
   {id: 'url', label: 'Foodbank Website'},
   {id: 'clientIntakeNumber', label: 'Client Intake Number'},
   {id: 'supportNumber', label: 'Support Number'},
+  {id: 'EIN', label: 'EIN Number'},
```
Comment (jspaine) [inline common/placeholders.js]: Make the id lower case to match the others. Think the label should just be 'EIN' too, or 'Employer Identification Number'

File: server/models/settings.js [hunk 2]
```diff
@@ -51,7 +51,7 @@ const SettingsSchema = new Schema({
   EIN: {
     type: String,
     trim: true,
-    default: ''
+    default: '0'
```
Comment (jspaine) [inline server/models/settings.js]: I'd just leave this blank - `default: ''`

File: common/placeholders.js [hunk 3]
```diff
@@ -76,6 +84,15 @@ function table(value) {
   return tableFinal + `<tr><td><p>Total:</p></td><td><p>"${total}"</p></td></tr></table>`
 }
 
+function tableReceipt(placeholders) {
+  let head = ['Foodbank Name','Foodbank Address','Foodbank Website','Client Intake Number', 'Support Number','EIN Number']
+  let tableFinal = `<table>`
+  for (let i = 0; i < 6; i++) {
+    tableFinal += `<tr><td><p>"${head[i]}":</p></td><td><p>"${placeholders[i]}"</p></td></tr>`
```
Comment (jspaine) [inline common/placeholders.js]: Think a better way to do this is something like:
```js
function renderReceipt(values) {
  const ids = ['organization', 'address', 'url', 'supportNumber', 'ein']
  let table = '<table>'

  for (let i = 0; i < ids.length; i++) {
    const placeholder = placeholders.find(p => p.id === ids[i])
    table += `<tr><td>${placeholder.label}: ${values[placeholder.id]}</td></tr>`
  }

  // then generate the items table
}
```
The `server/lib/mail/mail-helpers.js` `send` function already passes the settings fields so you shouldn't need to change the `bindPlaceholders` function in the mail generator.

File: server/models/settings.js [hunk 4]
```diff
@@ -51,7 +51,7 @@ const SettingsSchema = new Schema({
   EIN: {
```
Comment (jspaine) [inline server/models/settings.js]: Should also be lower case

**Comment by ronaldblanco** [comment]
Hello @jspaine; check now, I did left the ein default: '0' because the ein placeholder have problems with an empty string.

File: server/lib/mail/mail-generator.js [hunk 5]
```diff
@@ -74,9 +74,18 @@ function transformSubject(subject, bindings) {
 }
 
 function bindPlaceholder(node, bindings, attachments) {
+
   const id = getAttr(node, 'data-id')
   const placeholder = placeholders.find(pl => pl.id === id)
 
+  //Receipt Placeholder only
+  if (id === 'receipt') {
```
Comment (jspaine) [inline server/lib/mail/mail-generator.js:82]: Could probably just do:
```js
if (id === 'receipt') {
  bindings.receipt = {...bindings}
}
```

File: common/placeholders.js [hunk 6]
```diff
@@ -40,6 +41,13 @@ const placeholders = [
     required: true,
```
Comment (jspaine) [inline common/placeholders.js]: remember to delete this placeholder and the `table` function

Comment (ronaldblanco) [reply common/placeholders.js]: You do not think it could be use-full later??

File: server/lib/mail/mail-generator.js (hunk 5 above)
Comment (ronaldblanco) [reply server/lib/mail/mail-generator.js:82]: Yes, later I did think on that.

**Comment by ronaldblanco** [comment]
Check now and let me know!!

File: server/lib/mail/mail-generator.js (hunk 5 above)
Comment (jspaine) [reply server/lib/mail/mail-generator.js:82]: That's ok too, but you can just return `h(placeholder.format(bindings))` i think. `format` should be a function, and if it's not, it doesn't make much sense to return a text node of the bindings object anyway.

File: common/placeholders.js [hunk 7]
```diff
@@ -66,14 +67,25 @@ function getPlaceholders(types) {
...
+
+  table += `<tr><td><p>Here you have a list of the donations our organization recibed from you:</p></td></tr>`
+  const items = values.items
+  table += `<tr><td><p>Item</p></td><td><p>Value</p></td></tr>`
   let total = 0
-  for (let i = 0; i < value.length; i++) {
-    tableFinal += `<tr><td><p>"${value[i].name}"</p></td><td><p>"${value[i].value}"</p></td></tr>`
-    total += value[i].value
+  for (let i = 0; i < items.length; i++) {
+    table += `<tr><td><p>"${items[i].name}"</p></td><td><p>"${items[i].value}"</p></td></tr>`
```
Comment (jspaine) [inline common/placeholders.js]: Why the " 's around item name, value and total? Doesn't it look weird?

File: common/placeholders.js (hunk 6 above)
Comment (jspaine) [reply common/placeholders.js]: I don't see how. In any case, if you're not actually planning to use it for something, get rid of it. It'll still be in the git history.

**Comment by jspaine** [comment]
What was the problem with the ein field being an empty string by default?

**Comment by ronaldblanco** [comment]
When ein = "" the placeholder ein does not return, it give a problem of "No placeholder with that id" or something like that.
I thing that have to be with the mail-generator.js, some validation missing.

**Comment by ronaldblanco** [comment]
OK, I will be checking the rest you told me.

**Comment by jspaine** [comment]
Ah I think it's line 89 in the mail generator:
```js
if (!bindings[id] && placeholder.type !== placeholderTypes.ATTACHMENT)
```
should probably be:
```js
// import {isUndefined} from 'lodash'
if (isUndefined(bindings[id]) && placeholder.type !== placeholderTypes.ATTACHMENT)
```

File: common/placeholders.js (hunk 7 above)
Comment (ronaldblanco) [reply common/placeholders.js]: What are you referring exactly??
I do not understand!

**Comment by ronaldblanco** [comment]
You were right with the correction on line 89.
Check now and let me know.

File: common/placeholders.js (hunk 7 above)
Comment (jspaine) [reply common/placeholders.js]: Sorry! I don't think you need the double quotes, it should just show $15 for example, not "$15".

**Comment by ronaldblanco** [comment]
Ready.

**Comment by jspaine** [comment]
Thanks!

**Pull request merged**